File: Week 02/id_682/94.binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        white, gray = 0, 1
        res = []
        stack = [(white, root)]
        
        while stack:
            color, node = stack.pop()
            
            if node is None:
                continue
            
            if color == white:
                stack.append((white, node.right))
                stack.append((gray, node))
                stack.append((white, node.left))
            else:
                res.append(node.val)
                
        return res

File: Week 02/id_662/94.binary-tree-inorder-traversal.java
/**
 * 给定一个二叉树，返回它的中序 遍历。
 * <p>
 * 示例:
 * <p>
 * 输入: [1,null,2,3]
 * 1
 * \
 * 2
 * /
 * 3
 * <p>
 * 输出: [1,3,2]
 */
//递归
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        if(root==null){
            return list;
        }
        return inorderTraversalHelper(root,list);
    }

    public List<Integer> inorderTraversalHelper(TreeNode root,List<Integer> list) {
        if(root.left!=null){
            inorderTraversalHelper(root.left,list);
        }
        list.add(root.val);
        if(root.right!=null){
            inorderTraversalHelper(root.right,list);
        }
        return list;
    }
}

//迭代
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        while(root!=null || !stack.isEmpty()){
            if(root!=null){
                stack.push(root);
                root = root.left;
            }else{
                root = stack.pop();
                list.add(root.val);
                root = root.right;
            }
        }
        return list;
    }
}